linkis-computation-governance/linkis-manager/linkis-application-manager/src/main/java/org/apache/linkis/manager/am/manager/DefaultEngineNodeManager.java

```diff
@@ -0,0 +1,359 @@
+
+    nodes.forEach(
+        node -> {
+          Optional<NodeMetrics> nodeMetricsOptional =
+              Optional.ofNullable(metricses.get(node.getServiceInstance().toString()));
+          nodeMetricsOptional.ifPresent(m -> metricsConverter.fillMetricsToNode(node, m));
+        });
+    return nodes;
+  }
+
+  @Override
+  public EngineNode getEngineNodeInfo(EngineNode engineNode) {
+    /** 修改为实时请求对应的EngineNode */
+    EngineNodePointer engine = nodePointerBuilder.buildEngineNodePointer(engineNode);
+    NodeHeartbeatMsg heartMsg = engine.getNodeHeartbeatMsg();
+    engineNode.setNodeHealthyInfo(heartMsg.getHealthyInfo());
+    engineNode.setNodeOverLoadInfo(heartMsg.getOverLoadInfo());
+    engineNode.setNodeStatus(heartMsg.getStatus());
+    return engineNode;
+  }
+
+  @Override
+  public EngineNode getEngineNodeInfoByDB(EngineNode engineNode) {
+    // 1. 从持久化器中获取EngineNode信息，需要获取Task信息和Status信息，方便后面使用
+    engineNode = nodeManagerPersistence.getEngineNode(engineNode.getServiceInstance());
+    metricsConverter.fillMetricsToNode(
+        engineNode, nodeMetricManagerPersistence.getNodeMetrics(engineNode));
+    return engineNode;
+  }
+
+  @Override
+  public void updateEngineStatus(
+      ServiceInstance serviceInstance, NodeStatus fromState, NodeStatus toState) {}
+
+  @Override
+  public void updateEngine(EngineNode engineNode) {
+    nodeManagerPersistence.updateNodeInstance(engineNode);
+  }
+
+  @Override
+  public EngineNode switchEngine(EngineNode engineNode) {
+    return null;
+  }
+
+  @Override
+  public EngineNode reuseEngine(EngineNode engineNode) {
+    EngineNode node = getEngineNodeInfo(engineNode);
+    if (!NodeStatus.isAvailable(node.getNodeStatus())) {
+      return null;
+    }
+    if (!NodeStatus.isLocked(node.getNodeStatus())) {
+      Optional<String> lockStr =
+          engineLocker.lockEngine(node, (long) AMConfiguration.ENGINE_LOCKER_MAX_TIME.getValue());
+      if (!lockStr.isPresent()) {
+        throw new WarnException(
+            AMConstant.ENGINE_ERROR_CODE,
+            String.format(
+                "Failed to request lock from engine by reuse %s", node.getServiceInstance()));
+      }
+      node.setLock(lockStr.get());
+      return node;
+    } else {
+      return null;
+    }
+  }
+
+  /**
+   * TODO use Engine需要考虑流式引擎的场景，后续需要通过Label加额外的处理
+   *
+   * @param engineNode
+   * @param timeout
+   * @return
+   */
+  @Override
+  public EngineNode useEngine(EngineNode engineNode, long timeout) {
+    RetryHandler<EngineNode> retryHandler = new DefaultRetryHandler<EngineNode>();
+    retryHandler.addRetryException(feign.RetryableException.class);
+    retryHandler.addRetryException(UndeclaredThrowableException.class);
+
+    // wait until engine to be available
+    EngineNode node = retryHandler.retry(() -> getEngineNodeInfo(engineNode), "getEngineNodeInfo");
+    long retryEndTime = System.currentTimeMillis() + 60 * 1000;
```

Comment:
because spark engine connection start slow ,when first time call 'getEngineNodeInfo'  method, this method return engineNode with the NodeStatus equals to 'Starting' ,then useEngine(EngineNode engineNode, long timeout)  method  would return null, and LinkisRetryException exception would be thrown .
```
 EngineNode createEngineNode = getEngineNodeManager().useEngine(createNode, timeout);
              if (createEngineNode == null) {
                String message =
                    String.format(
                        "create engine%s success, but to use engine failed",
                        createNode.getServiceInstance());
                throw new LinkisRetryException(AMConstant.EM_ERROR_CODE, message);
              }
```